File: AccuPay/Domain/TimeEntry/TimeEntryGenerator.vb
Option Strict On

Imports System.Collections.ObjectModel
Imports System.Threading
Imports System.Threading.Tasks
Imports AccuPay.Entity
Imports AccuPay.Tools
Imports log4net
Imports Microsoft.EntityFrameworkCore
Imports Microsoft.Extensions.Logging.Console
Imports PayrollSys

Public Class TimeEntryGenerator
    Private Shared logger As ILog = LogManager.GetLogger("TimeEntryLogger")

    Private ReadOnly _cutoffStart As Date

    Private ReadOnly _cutoffEnd As Date
    Private ReadOnly _threeDays As Integer = 3
    Private _timeEntries As IList(Of TimeEntry)
    Private _actualTimeEntries As IList(Of ActualTimeEntry)
    Private _timeLogs As IList(Of TimeLog)
    Private _overtimes As IList(Of Overtime)
    Private _leaves As IList(Of Leave)
    Private _officialBusinesses As IList(Of OfficialBusiness)
    Private _agencyFees As IList(Of AgencyFee)
    Private _employeeShifts As IList(Of ShiftSchedule)
    Private _salaries As IList(Of Salary)
    Private _shiftSchedules As IList(Of EmployeeDutySchedule)
    Private _timeAttendanceLogs As List(Of TimeAttendanceLog)
    Private _breakTimeBrackets As List(Of BreakTimeBracket)

    Private _total As Integer

    Private _finished As Integer

    Private _errors As Integer
    Private _employees As IList(Of Employee)

    Public ReadOnly Property ErrorCount As Integer
        Get
            Return _errors
        End Get
    End Property

    Public ReadOnly Property Progress As Integer
        Get
            If _finished = 0 Then
                Return 0
            End If

            Return CInt(Math.Floor(_finished / _total * 100))
        End Get
    End Property

    Public Sub New(cutoffStart As Date, cutoffEnd As Date)
        _cutoffStart = cutoffStart
        _cutoffEnd = cutoffEnd
    End Sub

    Public Sub Start()
        Dim employees As IList(Of Employee) = Nothing
        Dim organization As Organization = Nothing
        Dim settings As ListOfValueCollection = Nothing
        Dim agencies As IList(Of Agency) = Nothing
        Dim calendarCollection As CalendarCollection

        Dim timeEntryPolicy As TimeEntryPolicy
        Using context = New PayrollContext()

            settings = New ListOfValueCollection(context.ListOfValues.ToList())
            timeEntryPolicy = New TimeEntryPolicy(settings)

            employees = context.Employees.
                Where(Function(e) e.OrganizationID.Value = z_OrganizationID).
                Include(Function(e) e.Position).
                ToList()
            _employees = employees

            agencies = context.Agencies.
                Where(Function(a) a.OrganizationID.Value = z_OrganizationID).
                ToList()

            organization = context.Organizations.
                SingleOrDefault(Function(o) o.RowID.Value = z_OrganizationID)

            _salaries = context.Salaries.
                Where(Function(s) s.OrganizationID.Value = z_OrganizationID).
                Where(Function(s) s.EffectiveFrom <= _cutoffStart).
                OrderByDescending(Function(s) s.EffectiveFrom).
                GroupBy(Function(s) s.EmployeeID).
                Select(Function(g) g.FirstOrDefault()).
                ToList()

            Dim previousCutoff = _cutoffStart.AddDays(-_threeDays)
            Dim endOfCutOff As Date = _cutoffEnd
            If timeEntryPolicy.PostLegalHolidayCheck Then endOfCutOff = _cutoffEnd.AddDays(_threeDays)
            _timeEntries = context.TimeEntries.
                Where(Function(t) t.OrganizationID.Value = z_OrganizationID).
                Where(Function(t) previousCutoff <= t.Date AndAlso t.Date <= endOfCutOff).
                ToList()

            _actualTimeEntries = context.ActualTimeEntries.
                Where(Function(a) a.OrganizationID.Value = z_OrganizationID).
                Where(Function(a) _cutoffStart <= a.Date AndAlso a.Date <= _cutoffEnd).
                ToList()

            _timeLogs = context.TimeLogs.
                Where(Function(t) t.OrganizationID.Value = z_OrganizationID).
                Where(Function(t) _cutoffStart <= t.LogDate AndAlso t.LogDate <= _cutoffEnd).
                ToList()

            _leaves = context.Leaves.
                Where(Function(l) l.OrganizationID.Value = z_OrganizationID).
                Where(Function(l) _cutoffStart <= l.StartDate AndAlso l.StartDate <= _cutoffEnd).
                Where(Function(l) l.Status = Leave.StatusApproved).
                ToList()

            _overtimes = context.Overtimes.
                Where(Function(o) o.OrganizationID.Value = z_OrganizationID).
                Where(Function(o) _cutoffStart <= o.OTStartDate AndAlso o.OTStartDate <= _cutoffEnd).
                Where(Function(o) o.Status = Overtime.StatusApproved).
                ToList()

            _officialBusinesses = context.OfficialBusinesses.
                Where(Function(o) o.OrganizationID.Value = z_OrganizationID).
                Where(Function(o) o.StartDate.Value <= _cutoffEnd AndAlso _cutoffStart <= o.EndDate.Value).
                Where(Function(o) o.Status = OfficialBusiness.StatusApproved).
                ToList()

            _agencyFees = context.AgencyFees.
                Where(Function(a) a.OrganizationID.Value = z_OrganizationID).
                Where(Function(a) _cutoffStart <= a.Date AndAlso a.Date <= _cutoffEnd).
                ToList()

            _employeeShifts = context.ShiftSchedules.
                Include(Function(s) s.Shift).
                Where(Function(s) s.OrganizationID = z_OrganizationID).
                Where(Function(s) s.EffectiveFrom <= _cutoffEnd AndAlso _cutoffStart <= s.EffectiveTo).
                ToList()

            _shiftSchedules = context.EmployeeDutySchedules.
                Where(Function(es) es.OrganizationID.Value = z_OrganizationID).
                Where(Function(es) _cutoffStart <= es.DateSched AndAlso es.DateSched <= _cutoffEnd).
                ToList()

            If timeEntryPolicy.ComputeBreakTimeLate Then
                _timeAttendanceLogs = context.TimeAttendanceLogs.
                                Where(Function(t) Nullable.Equals(t.OrganizationID, z_OrganizationID)).
                                Where(Function(t) _cutoffStart <= t.WorkDay AndAlso t.WorkDay <= _cutoffEnd).
                                ToList()

                _breakTimeBrackets = context.BreakTimeBrackets.
                                Include(Function(b) b.Division).
                                Where(Function(b) Nullable.Equals(b.Division.OrganizationID, z_OrganizationID)).
                                ToList()
            Else
                _timeAttendanceLogs = New List(Of TimeAttendanceLog)
                _breakTimeBrackets = New List(Of BreakTimeBracket)
            End If

            Dim payrateCalculationBasis = settings.GetEnum("Pay rate.CalculationBasis",
                                            AccuPay.PayRateCalculationBasis.Organization)

            If payrateCalculationBasis = AccuPay.PayRateCalculationBasis.Branch Then
                Dim branches = context.Branches.ToList()

                Dim calendarDays = context.CalendarDays.
                    Include(Function(t) t.DayType).
                    Where(Function(t) previousCutoff <= t.Date AndAlso t.Date <= _cutoffEnd).
                    ToList()

                calendarCollection = New CalendarCollection(branches, calendarDays)
            Else
                Dim payrates = context.PayRates.
                    Where(Function(p) p.OrganizationID.Value = z_OrganizationID).
                    Where(Function(p) previousCutoff <= p.Date AndAlso p.Date <= _cutoffEnd).
                    ToList()

                calendarCollection = New CalendarCollection(payrates)
            End If
        End Using

        Dim progress = New ObservableCollection(Of Integer)

        _total = employees.Count

        Parallel.ForEach(employees,
            Sub(employee)
                Try
                    CalculateEmployeeEntries(employee, organization, settings, agencies, timeEntryPolicy, calendarCollection)
                Catch ex As Exception
                    logger.Error(ex.Message, ex)
                    _errors += 1
                End Try

                Interlocked.Increment(_finished)
            End Sub)
    End Sub

    Private Sub CalculateEmployeeEntries(employee As Employee,
                                         organization As Organization,
                                         settings As ListOfValueCollection,
                                         agencies As IList(Of Agency),
                                         timeEntryPolicy As TimeEntryPolicy,
                                         calendarCollection As CalendarCollection)
        Dim previousTimeEntries As IList(Of TimeEntry) = _timeEntries.
            Where(Function(t) Nullable.Equals(t.EmployeeID, employee.RowID)).
            ToList()

        Dim actualTimeEntries As IList(Of ActualTimeEntry) = _actualTimeEntries.
            Where(Function(a) Nullable.Equals(a.EmployeeID, employee.RowID)).
            ToList()

        Dim salary = _salaries.
            FirstOrDefault(Function(s) Nullable.Equals(s.EmployeeID, employee.RowID))

        Dim timeLogs As IList(Of TimeLog) = _timeLogs.
            Where(Function(t) Nullable.Equals(t.EmployeeID, employee.RowID)).
            ToList()

        Dim shiftSchedules As IList(Of ShiftSchedule) = _employeeShifts.
            Where(Function(s) Nullable.Equals(s.EmployeeID, employee.RowID)).
            ToList()

        Dim overtimesInCutoff As IList(Of Overtime) = _overtimes.
            Where(Function(o) Nullable.Equals(o.EmployeeID, employee.RowID)).
            ToList()

        Dim officialBusinesses As IList(Of OfficialBusiness) = _officialBusinesses.
            Where(Function(o) Nullable.Equals(o.EmployeeID, employee.RowID)).
            ToList()

        Dim leavesInCutoff As IList(Of Leave) = _leaves.
            Where(Function(l) Nullable.Equals(l.EmployeeID, employee.RowID)).
            Where(Function(l) l.LeaveType <> "Leave w/o Pay").
            ToList()

        Dim agencyFees As IList(Of AgencyFee) = _agencyFees.
            Where(Function(a) Nullable.Equals(a.EmployeeID, employee.RowID)).
            ToList()

        Dim dutyShiftSchedules = _shiftSchedules.
            Where(Function(es) Nullable.Equals(es.EmployeeID, employee.RowID)).
            ToList()

        Dim timeAttendanceLogs As IList(Of TimeAttendanceLog) = _timeAttendanceLogs.
            Where(Function(t) Nullable.Equals(t.EmployeeID, employee.RowID)).
            ToList()

        Dim breakTimeBrackets As IList(Of BreakTimeBracket) = _breakTimeBrackets.
            Where(Function(b) Nullable.Equals(b.DivisionID, employee.Position?.DivisionID)).
            ToList()

        Dim payrateCalendar = calendarCollection.GetCalendar(employee)

        If employee.IsActive = False Then
            Dim currentTimeEntries = previousTimeEntries.
                Where(Function(t) _cutoffStart <= t.Date And t.Date <= _cutoffEnd)

            If Not currentTimeEntries.Any() Then
                Return
            End If
        End If

        If Not (timeLogs.Any() Or leavesInCutoff.Any() Or officialBusinesses.Any()) AndAlso (Not employee.IsFixed) Then
            Return
        End If

        Dim dayCalculator = New DayCalculator(organization, settings, payrateCalendar, employee)

        Dim timeEntries = New List(Of TimeEntry)
        For Each currentDate In Calendar.EachDay(_cutoffStart, _cutoffEnd)

            Try
                Dim timelog = timeLogs.OrderByDescending(Function(t) t.LastUpd).FirstOrDefault(Function(t) t.LogDate = currentDate)
                Dim employeeShift = shiftSchedules.FirstOrDefault(Function(s) s.EffectiveFrom <= currentDate And currentDate <= s.EffectiveTo)
                Dim overtimes = overtimesInCutoff.Where(Function(o) o.OTStartDate <= currentDate And currentDate <= o.OTEndDate).ToList()
                Dim leaves = leavesInCutoff.Where(Function(l) l.StartDate = currentDate).ToList()
                Dim officialBusiness = officialBusinesses.FirstOrDefault(Function(o) o.StartDate.Value = currentDate)
                Dim dutyShiftSched = dutyShiftSchedules.FirstOrDefault(Function(es) es.DateSched = currentDate)
                Dim currentTimeAttendanceLogs = timeAttendanceLogs.Where(Function(l) l.WorkDay = currentDate).ToList()

                Dim timeEntry = dayCalculator.Compute(
                    currentDate,
                    salary,
                    previousTimeEntries,
                    employeeShift,
                    dutyShiftSched,
                    timelog,
                    overtimes,
                    officialBusiness,
                    leaves,
                    currentTimeAttendanceLogs,
                    breakTimeBrackets)

                timeEntries.Add(timeEntry)
            Catch ex As Exception
                Throw New Exception($"{currentDate} #{employee.EmployeeNo}", ex)
            End Try

        Next

        PostLegalHolidayCheck(timeEntries, payrateCalendar, timeEntryPolicy)
        timeEntries.ForEach(Sub(t)
                                t.RegularHolidayPay += t.BasicRegularHolidayPay
                                t.TotalDayPay += t.BasicRegularHolidayPay
                            End Sub)

        If employee.IsUnderAgency Then
            Dim agency = agencies.SingleOrDefault(Function(a) Nullable.Equals(a.RowID, employee.AgencyID))

            Dim agencyCalculator = New AgencyFeeCalculator(employee, agency, agencyFees)
            agencyFees = agencyCalculator.Compute(timeEntries)
        End If

        Dim actualTimeEntryCalculator = New ActualTimeEntryCalculator(salary, actualTimeEntries, New ActualTimeEntryPolicy(settings))
        actualTimeEntries = actualTimeEntryCalculator.Compute(timeEntries)

        Using context = New PayrollContext()
            AddTimeEntriesToContext(context, timeEntries)
            AddActualTimeEntriesToContext(context, actualTimeEntries)
            AddAgencyFeesToContext(context, agencyFees)
            context.SaveChanges()
        End Using
    End Sub

    Private Sub PostLegalHolidayCheck(timeEntries As List(Of TimeEntry), payrateCalendar As PayratesCalendar, timeEntryPolicy As TimeEntryPolicy)
        If timeEntryPolicy.PostLegalHolidayCheck Then
            Dim employeeId = timeEntries.FirstOrDefault?.EmployeeID
            Dim employees = _employees.
                Where(Function(e) e.CalcHoliday).
                Where(Function(e) Equals(employeeId, e.RowID)).
                ToList()

            Dim isEntitledForLegalHolidayPay = employees.Any()
            If Not isEntitledForLegalHolidayPay Then Return

            Dim legalHolidays = payrateCalendar.LegalHolidays

            If legalHolidays.Any() Then
                Dim legalHolidayDescDates = legalHolidays.
                    Select(Function(p) p.Date).
                    OrderByDescending(Function(d) d).
                    ToList()

                Dim employee = employees.FirstOrDefault
                For Each holidayDate In legalHolidayDescDates
                    Dim presentAfterLegalHoliday = PayrollTools.HasWorkAfterLegalHoliday(
                        holidayDate, _cutoffEnd, timeEntries, payrateCalendar)

                    If Not presentAfterLegalHoliday Then
                        Dim timeEntry = timeEntries.Where(Function(t) t.Date = holidayDate).FirstOrDefault

                        timeEntry.BasicRegularHolidayPay = 0
                        'If employee.IsMonthly Then

                        'ElseIf employee.IsDaily Then

                        'End If
                    End If
                Next
            End If
        End If
    End Sub

    Private Sub AddTimeEntriesToContext(context As PayrollContext, timeEntries As IList(Of TimeEntry))
        For Each timeEntry In timeEntries
            If timeEntry.RowID.HasValue Then
                context.Entry(timeEntry).State = EntityState.Modified
            Else
                context.TimeEntries.Add(timeEntry)
            End If
        Next
    End Sub

    Private Sub AddActualTimeEntriesToContext(context As PayrollContext, actualTimeEntries As IList(Of ActualTimeEntry))
        For Each actualTimeEntry In actualTimeEntries
            If actualTimeEntry.RowID.HasValue Then
                context.Entry(actualTimeEntry).State = EntityState.Modified
            Else
                context.ActualTimeEntries.Add(actualTimeEntry)
            End If
        Next
    End Sub

    Private Sub AddAgencyFeesToContext(context As PayrollContext, agencyFees As IList(Of AgencyFee))
        For Each agencyFee In agencyFees
            If agencyFee.RowID.HasValue Then
                context.Entry(agencyFee).State = EntityState.Modified
            Else
                context.AgencyFees.Add(agencyFee)
            End If
        Next
    End Sub

    Private Function GetAgencyFees(context As PayrollContext, employee As Employee) As IList(Of AgencyFee)
        Return context.AgencyFees.
            Where(Function(a) Nullable.Equals(a.EmployeeID, employee.RowID)).
            Where(Function(a) _cutoffStart <= a.Date And a.Date <= _cutoffEnd).
            ToList()
    End Function

    Private Function GetSalary(context As PayrollContext, employee As Employee) As Salary
        Return context.Salaries.
            Where(Function(s) Nullable.Equals(s.EmployeeID, employee.RowID)).
            Where(Function(s) s.EffectiveFrom <= _cutoffStart And _cutoffStart <= If(s.EffectiveTo, _cutoffStart)).
            FirstOrDefault()
    End Function

    Private Function GetTimeEntries(context As PayrollContext, employee As Employee) As IList(Of TimeEntry)
        Dim previousCutoff = _cutoffStart.AddDays(-3)

        Return context.TimeEntries.
            Where(Function(t) Nullable.Equals(t.EmployeeID, employee.RowID)).
            Where(Function(t) previousCutoff <= t.Date And t.Date <= _cutoffEnd).
            ToList()
    End Function

    Private Function GetActualTimeEntries(context As PayrollContext, employee As Employee) As IList(Of ActualTimeEntry)
        Return context.ActualTimeEntries.
            Where(Function(t) Nullable.Equals(t.EmployeeID, employee.RowID)).
            Where(Function(t) _cutoffStart <= t.Date And t.Date <= _cutoffEnd).
            ToList()
    End Function

    Private Function GetTimeLogs(context As PayrollContext, employee As Employee) As IList(Of TimeLog)
        Return context.TimeLogs.
            Where(Function(t) Nullable.Equals(t.EmployeeID, employee.RowID)).
            Where(Function(t) _cutoffStart <= t.LogDate And t.LogDate <= _cutoffEnd).
            ToList()
    End Function

    Private Function GetShifts(context As PayrollContext, employee As Employee) As IList(Of ShiftSchedule)
        Return context.ShiftSchedules.
            Include(Function(s) s.Shift).
            Where(Function(s) Nullable.Equals(s.EmployeeID, employee.RowID)).
            Where(Function(s) s.EffectiveFrom <= _cutoffEnd And _cutoffStart <= s.EffectiveTo).
            ToList()
    End Function

    Private Function GetOvertimes(context As PayrollContext, employee As Employee) As IList(Of Overtime)
        Return context.Overtimes.
            Where(Function(o) Nullable.Equals(o.EmployeeID, employee.RowID)).
            Where(Function(o) o.OTStartDate <= _cutoffEnd And _cutoffStart <= o.OTEndDate).
            ToList()
    End Function

    Private Function GetOfficialBusinesses(context As PayrollContext, employee As Employee) As IList(Of OfficialBusiness)
        Return context.OfficialBusinesses.
            Where(Function(o) Nullable.Equals(o.EmployeeID, employee.RowID)).
            Where(Function(o) o.StartDate.Value <= _cutoffEnd And _cutoffStart <= o.EndDate.Value).
            Where(Function(o) o.Status = OfficialBusiness.StatusApproved).
            ToList()
    End Function

    Private Function GetLeaves(context As PayrollContext, employee As Employee) As IList(Of Leave)
        Return context.Leaves.
            Where(Function(l) Nullable.Equals(l.EmployeeID, employee.RowID)).
            Where(Function(l) _cutoffStart <= l.StartDate And l.StartDate <= _cutoffEnd).
            Where(Function(l) l.Status = Leave.StatusApproved).
            ToList()
    End Function

End Class

File: Accupay.Data/Services/TimEntryGeneration/TimeEntryGenerator.cs
using AccuPay.Data.Entities;
using AccuPay.Data.Enums;
using AccuPay.Data.Helpers;
using AccuPay.Data.Repositories;
using AccuPay.Data.ValueObjects;
using log4net;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace AccuPay.Data.Services
{
    public class TimeEntryGenerator
    {
        //private static ILog logger = LogManager.GetLogger("TimeEntryLogger", "TimeEntryLogger");

        private readonly int _organizationId;
        private readonly DateTime _cutoffStart;
        private readonly DateTime _cutoffEnd;
        private readonly int _threeDays = 3;
        private IList<TimeEntry> _timeEntries;
        private IList<ActualTimeEntry> _actualTimeEntries;
        private IList<TimeLog> _timeLogs;
        private IList<Overtime> _overtimes;
        private IList<Leave> _leaves;
        private IList<OfficialBusiness> _officialBusinesses;
        private IList<AgencyFee> _agencyFees;
        private IList<ShiftSchedule> _employeeShifts;
        private IList<Salary> _salaries;
        private IList<EmployeeDutySchedule> _shiftSchedules;
        private List<TimeAttendanceLog> _timeAttendanceLogs;
        private List<BreakTimeBracket> _breakTimeBrackets;

        private AgencyRepository _agencyRepository;
        private BreakTimeBracketRepository _breakTimeBracketRepository;
        private EmployeeRepository _employeeRepository;
        private EmployeeDutyScheduleRepository _employeeDutyScheduleRepository;
        private LeaveRepository _leaveRepository;
        private OfficialBusinessRepository _officialBusinessRepository;
        private OrganizationRepository _organizationRepository;
        private OvertimeRepository _overtimeRepository;
        private SalaryRepository _salaryRepository;
        private ShiftScheduleRepository _shiftScheduleRepository;
        private TimeLogRepository _timeLogRepository;
        private TimeAttendanceLogRepository _timeAttendanceLogRepository;

        private int _total;

        private int _finished;

        private int _errors;

        public int ErrorCount => _errors;

        public int Progress
        {
            get
            {
                if (_finished == 0)
                    return 0;

                return Convert.ToInt32(Math.Floor(_finished / (double)_total * 100));
            }
        }

        public TimeEntryGenerator(int organizationId, DateTime cutoffStart, DateTime cutoffEnd)
        {
            _organizationId = organizationId;
            _cutoffStart = cutoffStart;
            _cutoffEnd = cutoffEnd;

            _agencyRepository = new AgencyRepository();
            _breakTimeBracketRepository = new BreakTimeBracketRepository();
            _employeeRepository = new EmployeeRepository();
            _employeeDutyScheduleRepository = new EmployeeDutyScheduleRepository();
            _leaveRepository = new LeaveRepository();
            _officialBusinessRepository = new OfficialBusinessRepository();
            _organizationRepository = new OrganizationRepository();
            _overtimeRepository = new OvertimeRepository();
            _salaryRepository = new SalaryRepository();
            _shiftScheduleRepository = new ShiftScheduleRepository();
            _timeAttendanceLogRepository = new TimeAttendanceLogRepository();
            _timeLogRepository = new TimeLogRepository();
        }

        public void Start()
        {
            IList<Employee> employees = null;
            Organization organization = null;
            IList<Agency> agencies = null;
            CalendarCollection calendarCollection;

            ListOfValueCollection settings = ListOfValueCollection.Create();
            TimeEntryPolicy timeEntryPolicy = new TimeEntryPolicy(settings);

            TimePeriod cuttOffPeriod = new TimePeriod(_cutoffStart, _cutoffEnd);

            using (var context = new PayrollContext())
            {
                employees = _employeeRepository.GetAllActiveWithPosition(_organizationId).ToList();

                agencies = _agencyRepository.GetAll(_organizationId).ToList();

                organization = _organizationRepository.GetById(_organizationId);

                _salaries = _salaryRepository.GetByCutOff(_organizationId, _cutoffStart).ToList();

                var previousCutoff = PayrollTools.GetPreviousCutoffDateForCheckingLastWorkingDay(_cutoffStart);

                DateTime endOfCutOff = _cutoffEnd;
                if (timeEntryPolicy.PostLegalHolidayCheck)
                {
                    endOfCutOff = _cutoffEnd.AddDays(_threeDays);
                }

                // TODO: move this to a repository
                _timeEntries = context.TimeEntries.
                                    Where(t => t.OrganizationID.Value == _organizationId).
                                    Where(t => previousCutoff <= t.Date && t.Date <= endOfCutOff).
                                    ToList();

                // TODO: move this to a repository
                _actualTimeEntries = context.ActualTimeEntries.
                                    Where(a => a.OrganizationID.Value == _organizationId).
                                    Where(a => _cutoffStart <= a.Date && a.Date <= _cutoffEnd).
                                    ToList();

                _timeLogs = _timeLogRepository.
                                GetByDatePeriod(_organizationId, cuttOffPeriod).
                                ToList();

                _leaves = _leaveRepository.
                                GetAllApprovedByDatePeriod(_organizationId, cuttOffPeriod).
                                ToList();

                _overtimes = _overtimeRepository.
                                GetByDatePeriod(_organizationId, cuttOffPeriod, OvertimeStatus.Approved).
                                ToList();

                _officialBusinesses = _officialBusinessRepository.
                                GetAllApprovedByDatePeriod(_organizationId, cuttOffPeriod).
                                ToList();

                // TODO: move this to a repository
                _agencyFees = context.AgencyFees.
                                Where(a => a.OrganizationID.Value == _organizationId).
                                Where(a => _cutoffStart <= a.Date && a.Date <= _cutoffEnd).
                                ToList();

                _employeeShifts = _shiftScheduleRepository.
                                GetByDatePeriod(_organizationId, cuttOffPeriod).
                                ToList();

                _shiftSchedules = _employeeDutyScheduleRepository.
                                GetByDatePeriod(_organizationId, cuttOffPeriod).
                                ToList();

                if (timeEntryPolicy.ComputeBreakTimeLate)
                {
                    _timeAttendanceLogs = _timeAttendanceLogRepository.
                                            GetByTimePeriod(_organizationId, cuttOffPeriod).
                                            ToList();

                    _breakTimeBrackets = _breakTimeBracketRepository.
                                            GetAll(_organizationId).
                                            ToList();
                }
                else
                {
                    _timeAttendanceLogs = new List<TimeAttendanceLog>();
                    _breakTimeBrackets = new List<BreakTimeBracket>();
                }

                var payrateCalculationBasis = settings.GetEnum("Pay rate.CalculationBasis", PayRateCalculationBasis.Organization);

                calendarCollection = PayrollTools.GetCalendarCollection(
                                                        new TimePeriod(previousCutoff, _cutoffEnd),
                                                        payrateCalculationBasis,
                                                        _organizationId);
            }

            var progress = new ObservableCollection<int>();

            _total = employees.Count;

            Parallel.ForEach(employees, employee =>
            {
                try
                {
                    CalculateEmployeeEntries(employee, organization, settings, agencies, timeEntryPolicy, calendarCollection);
                }
                catch (Exception)
                {
                    // can error if employee type is null
                    //logger.Error(ex.Message, ex);
                    _errors += 1;
                }

                Interlocked.Increment(ref _finished);
            });
        }

        private void CalculateEmployeeEntries(Employee employee,
                                            Organization organization,
                                            ListOfValueCollection settings,
                                            IList<Agency> agencies,
                                            TimeEntryPolicy timeEntryPolicy,
                                            CalendarCollection calendarCollection)
        {
            IList<TimeEntry> previousTimeEntries = _timeEntries.
                                                        Where(t => t.EmployeeID == employee.RowID).
                                                        ToList();

            IList<ActualTimeEntry> actualTimeEntries = _actualTimeEntries.
                                                        Where(a => a.EmployeeID == employee.RowID).
                                                        ToList();

            var salary = _salaries.FirstOrDefault(s => s.EmployeeID == employee.RowID);

            IList<TimeLog> timeLogs = _timeLogs.
                                        Where(t => t.EmployeeID == employee.RowID).
                                        ToList();

            IList<ShiftSchedule> shiftSchedules = _employeeShifts.
                                        Where(s => s.EmployeeID == employee.RowID).
                                        ToList();

            IList<Overtime> overtimesInCutoff = _overtimes.
                                        Where(o => o.EmployeeID == employee.RowID).
                                        ToList();

            IList<OfficialBusiness> officialBusinesses = _officialBusinesses.
                                        Where(o => o.EmployeeID == employee.RowID).
                                        ToList();

            IList<Leave> leavesInCutoff = _leaves.Where(l => l.EmployeeID == employee.RowID).
                                        Where(l => l.LeaveType != "Leave w/o Pay").
                                        ToList();

            IList<AgencyFee> agencyFees = _agencyFees.
                                        Where(a => a.EmployeeID == employee.RowID).
                                        ToList();

            var dutyShiftSchedules = _shiftSchedules.
                                        Where(es => es.EmployeeID == employee.RowID).
                                        ToList();

            IList<TimeAttendanceLog> timeAttendanceLogs = _timeAttendanceLogs.
                                        Where(t => t.EmployeeID == employee.RowID).
                                        ToList();

            IList<BreakTimeBracket> breakTimeBrackets = _breakTimeBrackets.
                                        Where(b => b.DivisionID == employee.Position?.DivisionID).
                                        ToList();

            if (employee.IsActive == false)
            {
                var currentTimeEntries = previousTimeEntries.
                                        Where(t => _cutoffStart <= t.Date && t.Date <= _cutoffEnd);

                // TODO: return this as one the list of errors of Time entry generation
                if (!currentTimeEntries.Any())
                    return;
            }

            // TODO: return this as one the list of errors of Time entry generation
            if (!(timeLogs.Any() || leavesInCutoff.Any() || officialBusinesses.Any()) && (!employee.IsFixed))
                return;

            var dayCalculator = new DayCalculator(organization, settings, employee);

            var timeEntries = new List<TimeEntry>();
            var regularHolidaysList = new List<DateTime>(); // Used for postlegalholidaycheck
            foreach (var currentDate in CalendarHelper.EachDay(_cutoffStart, _cutoffEnd))
            {
                try
                {
                    var timelog = timeLogs.OrderByDescending(t => t.LastUpd).FirstOrDefault(t => t.LogDate == currentDate);
                    var employeeShift = shiftSchedules.FirstOrDefault(s => s.EffectiveFrom <= currentDate & currentDate <= s.EffectiveTo);
                    var overtimes = overtimesInCutoff.Where(o => o.OTStartDate <= currentDate & currentDate <= o.OTEndDate).ToList();
                    var leaves = leavesInCutoff.Where(l => l.StartDate == currentDate).ToList();
                    var officialBusiness = officialBusinesses.FirstOrDefault(o => o.StartDate.Value == currentDate);
                    var dutyShiftSched = dutyShiftSchedules.FirstOrDefault(es => es.DateSched == currentDate);
                    var currentTimeAttendanceLogs = timeAttendanceLogs.Where(l => l.WorkDay == currentDate).ToList();

                    var branchId = timelog?.BranchID ?? employee?.BranchID;
                    var payrate = calendarCollection.GetCalendar(branchId).Find(currentDate);

                    var timeEntry = dayCalculator.Compute(currentDate,
                                                            salary,
                                                            previousTimeEntries,
                                                            employeeShift,
                                                            dutyShiftSched,
                                                            timelog,
                                                            overtimes,
                                                            officialBusiness,
                                                            leaves,
                                                            currentTimeAttendanceLogs,
                                                            breakTimeBrackets,
                                                            payrate,
                                                            calendarCollection,
                                                            branchId);

                    if (payrate.IsRegularHoliday)
                    {
                        regularHolidaysList.Add(currentDate);
                    }

                    timeEntries.Add(timeEntry);
                }
                catch (Exception ex)
                {
                    throw new Exception($"{currentDate} #{employee.EmployeeNo}", ex);
                }
            }

            PostLegalHolidayCheck(employee, timeEntries, timeEntryPolicy, regularHolidaysList, calendarCollection);
            timeEntries.ForEach(t =>
            {
                t.RegularHolidayPay += t.BasicRegularHolidayPay;
                t.TotalDayPay += t.BasicRegularHolidayPay;
            });

            if (employee.IsUnderAgency)
            {
                var agency = agencies.SingleOrDefault(a => a.RowID == employee.AgencyID);

                var agencyCalculator = new AgencyFeeCalculator(employee, agency, agencyFees);
                agencyFees = agencyCalculator.Compute(timeEntries);
            }

            var actualTimeEntryCalculator = new ActualTimeEntryCalculator(salary, actualTimeEntries, new ActualTimeEntryPolicy(settings));
            actualTimeEntries = actualTimeEntryCalculator.Compute(timeEntries);

            using (var context = new PayrollContext())
            {
                AddTimeEntriesToContext(context, timeEntries);
                AddActualTimeEntriesToContext(context, actualTimeEntries);
                AddAgencyFeesToContext(context, agencyFees);
                context.SaveChanges();
            }
        }

        private void PostLegalHolidayCheck(Employee employee,
                                        List<TimeEntry> timeEntries,
                                        TimeEntryPolicy timeEntryPolicy,
                                        List<DateTime> regularHolidaysList,
                                        CalendarCollection calendarCollection)
        {
            if (timeEntryPolicy.PostLegalHolidayCheck)
            {
                if (!employee.CalcHoliday)
                    return;

                if (regularHolidaysList.Any())
                {
                    foreach (var holidayDate in regularHolidaysList)
                    {
                        var presentAfterLegalHoliday = PayrollTools.
                                                        HasWorkAfterLegalHoliday(holidayDate,
                                                                                _cutoffEnd,
                                                                                timeEntries,
                                                                                calendarCollection);

                        if (!presentAfterLegalHoliday)
                        {
                            var timeEntry = timeEntries.
                                        Where(t => t.Date == holidayDate).
                                        FirstOrDefault();

                            timeEntry.BasicRegularHolidayPay = 0;
                        }
                    }
                }
            }
        }

        private void AddTimeEntriesToContext(PayrollContext context, IList<TimeEntry> timeEntries)
        {
            foreach (var timeEntry in timeEntries)
            {
                if (timeEntry.RowID.HasValue)
                    context.Entry(timeEntry).State = EntityState.Modified;
                else
                    context.TimeEntries.Add(timeEntry);
            }
        }

        private void AddActualTimeEntriesToContext(PayrollContext context, IList<ActualTimeEntry> actualTimeEntries)
        {
            foreach (var actualTimeEntry in actualTimeEntries)
            {
                if (actualTimeEntry.RowID.HasValue)
                    context.Entry(actualTimeEntry).State = EntityState.Modified;
                else
                    context.ActualTimeEntries.Add(actualTimeEntry);
            }
        }

        private void AddAgencyFeesToContext(PayrollContext context, IList<AgencyFee> agencyFees)
        {
            foreach (var agencyFee in agencyFees)
            {
                if (agencyFee.RowID.HasValue)
                    context.Entry(agencyFee).State = EntityState.Modified;
                else
                    context.AgencyFees.Add(agencyFee);
            }
        }
    }
}